### PR TITLE
IGNITE-3303 Apache Flink Integration - Flink source to run a continuous query against one or multiple caches

### PR DESCRIPTION
**Pull request by samaitra**
Apache Flink integration
+++ Ignite as a bidirectional Connector +++

As a Flink source => run a continuous query against one or multiple caches.

Related discussion : http://apache-ignite-developers.2346864.n4.nabble.com/Apache-Flink-lt-gt-Apache-Ignite-integration-td8163.html

### CODE REVIEW

File: modules/flink/src/main/java/org/apache/ignite/source/flink/TaskRemoteFilter.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+        this.filter = filter;
+    }
+
+    /** {@inheritDoc} */
+    @Override public boolean apply(CacheEvent evt) {
+        Affinity<Object> affinity = ignite.affinity(cacheName);
+
+        // Process this event. Ignored on backups.
+        return affinity.isPrimary(ignite.cluster().localNode(), evt.key()) &&
+                !(filter != null && filter.apply(evt));
```
Comment (AMashenkov) [inline modules/flink/src/main/java/org/apache/ignite/source/flink/TaskRemoteFilter.java]: Why delegate result is negated here?
Looks like correct statement should be:
 affinity.isPrimary(ignite.cluster().localNode(), evt.key()) && **(filter == null || filter.apply(evt))**

File: modules/flink/src/main/java/org/apache/ignite/source/flink/IgniteSource.java [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+        stopped = true;
+
+        if (rmtLsnrId != null && ignite != null) {
+            ignite.events(ignite.cluster().forCacheNodes(cacheName))
+                .stopRemoteListen(rmtLsnrId);
+
+            rmtLsnrId = null;
+
+            ignite.cache(cacheName).close();
```
Comment (AMashenkov) [inline modules/flink/src/main/java/org/apache/ignite/source/flink/IgniteSource.java]: I'm not sure it is correct as IgniteSource uses a cache that was created outside of IgniteSource instance and can be used by someone else.

**Comment by samaitra** [comment]
@AMashenkov Thank you, for reviewing the changes. I have incorporated the changes.

File: modules/flink/src/main/java/org/apache/ignite/source/flink/TaskRemoteFilter.java [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.ignite.cache.affinity.Affinity;
+import org.apache.ignite.events.CacheEvent;
+import org.apache.ignite.lang.IgnitePredicate;
+import org.apache.ignite.resources.IgniteInstanceResource;
+
+/**
+ * Remote filter.
+ */
+public class TaskRemoteFilter implements IgnitePredicate<CacheEvent> {
+    /** Ignite Instance Resource. */
```
Comment (AMashenkov) [inline modules/flink/src/main/java/org/apache/ignite/source/flink/TaskRemoteFilter.java]: Let's add  some "private static final long serialVersionUID =" as class is serializable and designed to execute remotly.

File: modules/flink/src/main/java/org/apache/ignite/source/flink/IgniteSource.java [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    private static final int DFLT_EVT_BUFFER_TIMEOUT = 10;
+
+    /** Event buffer. */
+    private BlockingQueue<CacheEvent> evtBuf = new LinkedBlockingQueue<>();
+
+    /** Remote Listener id. */
+    private UUID rmtLsnrId;
+
+    /** Flag for stopped state. */
+    public AtomicBoolean stopped = new AtomicBoolean(false);
```
Comment (AMashenkov) [inline modules/flink/src/main/java/org/apache/ignite/source/flink/IgniteSource.java]: Why we can't just make this final and omit 'null' checks?

**Comment by dspavlov** [comment]
The ticket is resolved, closing PR. 

Please close PRs for your resolved tickets and other non-needed PRs. 

Alternatively, ask a committer 
- to use GitHub UI (committer  can associate accounts here: https://gitbox.apache.org/setup/ )
- or /scripts/apply_pull_request.sh from Apache Ignite source.
- or mention in commit message 'Fixes #(a number of PR)'. 

These methods will automatically close PR.

Feel free to reopen this PR if it is actual.

**Pull request closed without merging**